### DISCUSSION TITLE
Build errors, probably around emsdk

### DISCUSSION THREAD

**Discussion by julianna-ciq** [perspective-dev/perspective#2155] [Q&A] [answered]
I was looking to build the docs to fix [an issue I raised](https://github.com/finos/perspective/issues/2133), but I'm struggling to even build Perspective.

I'm on Windows 10, and I'm running all of my commands from Bash. I've installed CMake, Boost, and wasm-pack (+ rustup). I successfully installed all of the dependencies with `npm install`. I ran `npm run setup` and selected "Javascript" (so my .perspectiverc file says `PSP_PROJECT=js`).

That said, when I ran install, I did get this message:
> \> postinstall
> \> node scripts/install_emsdk.js
> '.' is not recognized as an internal or external command, operable program or batch file.

(I've confirmed that the error is thrown from [this line](https://github.com/finos/perspective/blob/master/scripts/install_emsdk.js#L48).)

...which is probably related to the fact that when I go run `npm run build`, I get this:

> $ npm run build
> 
> \> build
> \> node scripts/build.js
> 
> -- Running build for @finos/perspective-cpp,@finos/perspective-esbuild-plugin
> 
> \> @finos/perspective-cpp@1.9.4 build
> \> node ../../scripts/run_emsdk.js node ./build.js
> 
> The system cannot find the path specified.
> Command failed: cd C:\Users\Julianna\p2\.emsdk && . ./emsdk_env.sh >/dev/null 2>&1 && cd C:\Users\Julianna\p2\cpp\perspective && node ./build.js
> npm ERR! Lifecycle script `build` failed with error:
> npm ERR! Error: command failed
> npm ERR!   in workspace: @finos/perspective-cpp@1.9.4
> npm ERR!   at location: C:\Users\Julianna\p2\cpp\perspective
> 
> Command failed: npm --workspace @finos/perspective-cpp run build

Any advice on how to get unstuck from this?

**Comment by sinistersnare** [comment, marked as answer]
You can install and source `emsdk` by hand, and pass `PSP_SKIP_EMSDK_INSTALL=true` in your environment, so that it doesnt execute this file.

The issue is the `. ./emsdk_env.sh` because the windows command lines (idk which you are using) do not support `.` as a way to source files it seems.

**Reply by julianna-ciq** [reply]
I installed and built emsdk, which included adding emsdk to my PATH environment variable.

My .perspectiverc file looks like:
```
PSP_PROJECT=js
PSP_SKIP_EMSDK_INSTALL=true
```

At this point, I was able to successfully run `yarn install`.

I re-ran `yarn build` in git bash, and wound up with this error:
```
$ yarn build
yarn run v1.22.1
$ node scripts/build.js
-- Running build for @finos/perspective-cpp,@finos/perspective-esbuild-plugin

> @finos/perspective-cpp@2.0.0 build
> node ../../scripts/run_emsdk.js node ./build.js

The system cannot find the path specified.
Command failed: cd ~\.emsdk && . ./emsdk_env.sh >/dev/null 2>&1 && cd ~\cpp\perspective && node ./build.js
npm ERR! Lifecycle script `build` failed with error:
npm ERR! Error: command failed
npm ERR!   in workspace: @finos/perspective-cpp@2.0.0
npm ERR!   at location: ~\cpp\perspective

Command failed: npm --workspace @finos/perspective-cpp run build
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

This feels like a related error, but I'm not sure how to get around it without changing the build scripts.

**Reply by sinistersnare** [reply]
The problem is in `run_emsdk.js`. Are you running this (and is this repo's files) directly in your home directory? It is looking for `~\cpp\perspective` which seems... unlikely to exist. It probably exists closer to `~\perspective\cpp\perspective`.

So you probably need to run it in the correct directory, or even have your files in the correct directory.

The installation process should be something like:

```sh
git clone https://github.com/finos/perspective
cd perspective
yarn
# Do this echo in lieu of `yarn setup`
echo 'PSP_PROJECT=js\nPSP_SKIP_EMSDK_INSTALL=true\n` > .perspectiverc
yarn build
```

Although you gotta have a `.emsdk` directory in the perspective repo, cause it will look for it.

**Reply by julianna-ciq** [reply]
Sorry, I was just trying to cut out unnecessary parts of my file path, but I forgot that `~` meant something in other OS's.

I tried to restart from a clean slate. I ran these steps:

```
git clone https://github.com/finos/perspective
cd perspective
yarn set version 1.22.1
yarn
echo 'PSP_PROJECT=js\nPSP_SKIP_EMSDK_INSTALL=true\n' > .perspectiverc
yarn build
```
(my default yarn version is 3.4.2, which has its own set of issues, so I reset it to yarn's 1.x line)

After running that, I have the same error message:
```
$ yarn build
yarn run v1.22.1
$ node scripts/build.js
-- Running build for @finos/perspective-cpp,@finos/perspective-esbuild-plugin

> @finos/perspective-cpp@2.0.0 build
> node ../../scripts/run_emsdk.js node ./build.js

The system cannot find the path specified.
Command failed: cd C:\Users\Julianna\perspective\.emsdk && . ./emsdk_env.sh >/dev/null 2>&1 && cd C:\Users\Julianna\perspective\cpp\perspective && node ./build.js
npm ERR! Lifecycle script `build` failed with error:
npm ERR! Error: command failed
npm ERR!   in workspace: @finos/perspective-cpp@2.0.0
npm ERR!   at location: C:\Users\Julianna\perspective\cpp\perspective

Command failed: npm --workspace @finos/perspective-cpp run build
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

**Reply by sinistersnare** [reply]
You should run each command in 

> cd C:\Users\Julianna\perspective\.emsdk && . ./emsdk_env.sh >/dev/null 2>&1 && cd C:\Users\Julianna\perspective\cpp\perspective && node ./build.js

And figure which one is failing exactly. If you havent installed the emsdk, and there is no `.emsdk` directory, you need to install it.
